spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/MappingMongoConverter.java

```diff
@@ -1667,4 +1689,91 @@ private static boolean isCollectionOfDbRefWhereBulkFetchIsPossible(Iterable<?> s
 			return null;
 		}
 	}
+
+	private static class HijackedTypeInformation<S> implements TypeInformation<S> {
```

Comment:
How about `TypeInformationWrapper` 😬and handling the rest of customizations in a separated, smaller type?